### PR TITLE
Add IPAM to API

### PR DESCRIPTION
**Pull request by tmjd**
## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/apis/operator/v1/types.go [hunk 1]
```diff
@@ -369,18 +369,67 @@ func (cp CNIPluginType) String() string {
...
+	IPAMPluginAzureVNET.String(),
+}
+
+func (cp IPAMPluginType) String() string {
+	return string(cp)
+}
+
+type IPAMSpec struct {
+	// Specifies the IPAM plugin that will be used in the Calico or Calico Enterprise installation.
+	// * For CNI Plugin Calico, this field defaults to Calico.
```
Comment (tmjd) [inline pkg/apis/operator/v1/types.go:401]: This generates the following
```
<td>
<code>type</code><br>
<em>
<a href="#operator.tigera.io/v1.IPAMPluginType">
IPAMPluginType
</a>
</em>
</td>
<td>
<p>Specifies the IPAM plugin that will be used in the Calico or Calico Enterprise installation.</p>
<ul>
<li><p>For CNI Plugin Calico, this field defaults to Calico.</p></li>
<li><p>For CNI Plugin GKE, this field defaults to HostLocal.</p></li>
<li><p>For CNI Plugin AzureVNET, this field defaults to AzureVNET.</p></li>
<li><p>For CNI Plugin AmazonVPC, this field defaults to AmazonVPC.</p></li>
</ul>
<p>The IPAM plugin is installed and configured only if the CNI plugin is set to Calico,
for all other values of the CNI plugin the plugin binaries and CNI config is a dependency
that is expected to be installed separately.</p>
<p>Default: Calico</p>
</td>
```
So it does what I was hoping though the CRD description looks a little funky.

File: pkg/apis/operator/v1/types.go [hunk 2]
```diff
@@ -369,18 +369,67 @@ func (cp CNIPluginType) String() string {
...
+	IPAMPluginAmazonVPC.String(),
+	IPAMPluginAzureVNET.String(),
+}
+
+func (cp IPAMPluginType) String() string {
+	return string(cp)
+}
+
+type IPAMSpec struct {
+	// Specifies the IPAM plugin that will be used in the Calico or Calico Enterprise installation.
```
Comment (caseydavenport) [inline pkg/apis/operator/v1/types.go:400]: Shouldn't this comment be outside of the struct?

File: pkg/apis/operator/v1/types.go [hunk 3]
```diff
@@ -369,18 +369,67 @@ func (cp CNIPluginType) String() string {
...
+	//
 	// For the value Calico, the CNI plugin binaries and CNI config will be installed as part of deployment,
 	// for all other values the CNI plugin binaries and CNI config is a dependency that is expected
 	// to be installed separately.
 	//
 	// Default: Calico
 	// +kubebuilder:validation:Enum=Calico;GKE;AmazonVPC;AzureVNET
 	Type CNIPluginType `json:"type"`
+
+	// IPAM specifies the IPAM that will be used in the Calico or Calico Enterprise installation.
```
Comment (caseydavenport) [inline pkg/apis/operator/v1/types.go:430]: This is a bit self-referential (IPAM specifies the IPAM).

Maybe "IPAM configures pod IP address management"

File: pkg/render/node_test.go [hunk 4]
```diff
@@ -807,6 +810,74 @@ var _ = Describe("Node rendering tests", func() {
 		Expect(ds.Spec.Template.Spec.Containers[0].LivenessProbe).To(Equal(expectedLiveness))
 	})
 
+	DescribeTable("should properly render configuration using non-Calico CNI plugin",
+		func(cni operator.CNIPluginType, ipam operator.IPAMPluginType, expectedEnvs []v1.EnvVar) {
+			amazonVPCInstalllation := &operator.Installation{
```
Comment (caseydavenport) [inline pkg/render/node_test.go:815]: amazonVPCInstallation->installation

**Pull request merged**